pkg/db/queries.sql

```diff
@@ -13,23 +13,36 @@ WHERE
 	singleton_id = 1;
 
 -- name: InsertGatewayEnvelope :execrows
-SELECT
-	insert_gateway_envelope(@originator_node_id, @originator_sequence_id, @topic, @originator_envelope);
+INSERT INTO gateway_envelopes(originator_node_id, originator_sequence_id, topic, originator_envelope)
+	VALUES (@originator_node_id, @originator_sequence_id, @topic, @originator_envelope)
+ON CONFLICT
+	DO NOTHING;
 
 -- name: SelectGatewayEnvelopes :many
+WITH cursors AS (
+	SELECT
+		UNNEST(@cursor_node_ids::INT[]) AS cursor_node_id,
+		UNNEST(@cursor_sequence_ids::BIGINT[]) AS cursor_sequence_id
+)
 SELECT
-	*
+	gateway_envelopes.*
 FROM
 	gateway_envelopes
+	-- Assumption: There is only one cursor per node ID. Caller must verify this
+	LEFT JOIN cursors ON gateway_envelopes.originator_node_id = cursors.cursor_node_id
 WHERE (sqlc.narg('topic')::BYTEA IS NULL
 	OR length(@topic) = 0
 	OR topic = @topic)
 AND (sqlc.narg('originator_node_id')::INT IS NULL
 	OR originator_node_id = @originator_node_id)
-AND (sqlc.narg('originator_sequence_id')::BIGINT IS NULL
-	OR originator_sequence_id > @originator_sequence_id)
-AND (sqlc.narg('gateway_sequence_id')::BIGINT IS NULL
-	OR id > @gateway_sequence_id)
+AND (cursor_sequence_id IS NULL
+	OR originator_sequence_id > cursor_sequence_id)
+ORDER BY
+	-- Assumption: envelopes are inserted in sequence_id order per originator, therefore
+	-- gateway_time preserves sequence_id order
+	gateway_time,
```

Comment:
This sorts the envelopes roughly in the order they were received. The reason this should match sequence_id order is that:
- For envelopes originated elsewhere, we are inserting them sequentially from the stream
- For envelopes originated by ourselves, we are inserting them sequentially via the publish worker

There are some alternative sort methods I considered:
- Sort by originator time. The advantage is the ordering is the same on all gateways.
- Sort by originator node ID/originator sequence ID.

Sorting by gateway_time has the advantage that clients receive payloads in the same order from the gateway regardless of the time that they chose to sync (you cannot receive a payload later on that would have been inserted earlier in the sort order). Note that if we want to sort by originator timestamp in the case of causal ordering tiebreaks, clients can still do that locally. I can definitely be convinced otherwise on this decision though.